Sources/LambdaMocks/ContextProvider.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  ContextProvider.swift
+//  LambdaExtras
+//
+//  Created by Mathew Gacy on 1/7/24.
+//
+
+import Foundation
+import Logging
+import NIOCore
+import NIO
+
+/// A helper to create and manage mock initialization and runtime contexts for testing.
+///
+/// Example usage:
+///
+/// ```swift
+/// final class MyHandlerTests: XCTestCase {
+///     var contextProvider: ContextProvider<MyEnvironment>
+///
+///     override func setUp() {
+///         contextProvider.setUp()
+///     }
+///
+///     override func tearDown() {
+///         XCTAssertNoThrow(try contextProvider.shutdown())
+///     }
+///
+///     func testMyHandler() async throws {
+///         let sut = try await MyHandler(context: contextProvider.makeInitializationContext())
+///         let actual = try await sut.handle(MockEvent(), context: contextProvider.makeContext())
+///         ...
+///     }
+/// }
+/// ```
+public struct ContextProvider<E> {
```

Comment:
Good point, especially when I have used `EnvironmentVariable` elsewhere

Reply:
Looks like it is `E` on `MockContext` as well 🤷‍♂️

Reply:
Updated `MockContext` and `MockInitializationContext`

Reply:
This does bring up some confusion on my part around why these types are generic. I realize this relates to code that is already approved/merged, however `Lambda.env(_:)` takes a concrete `String` type. Could you clarify the added flexibility?

Reply:
The basic idea is types > strings

Let's say I define 2 environment variables for my Lambda in my SAM template:

- `bar`
- `baz`

Using `Lambda.env(_:)` the compiler would allow:

```swift
guard let bar = Lambda.env("barrr") else {
    throw HandlerError.envError("bar")
}
```

With this I can do:

```swift
enum Environment: String {
    /// Explanation of `bar`
    case bar
    /// Explanation of `baz`
    case baz
}

...

let bar = try context.value(for: .baz)
```

While I do offer a `LambdaExtras.DefaultEnvironment` type, in most cases there will be a project-specific `Environment` type defined in a target using this package. The generics allow the contexts and other types to support those externally-defined types.

Reply:
So, there are other ways to get around the issues I pointed to above, but more generally, this setup means that instead of:

```swift
import AWSLambdaRuntimeCore

struct Foo {
    var stuff: (LambdaContext) async throws -> String
}

extension Foo: DependencyKey {
    static var liveValue = Foo(
        // To control environment variables we need to either export them before calling
        // `swift test` or access them inside handler dependencies that we'll mock
        // in tests
        stuff: { context in
            let bar = try Lambda.env(.bar)
                .unwrap(or: "Could not find bar in env vars.")
            
            ... other stuff having nothing to do with `context`
        }
    )
}

public struct MyHandler {
    @Dependency(\.foo) var foo
    
    public func run(with event: Event, context: LambdaContext) async throws {
        let stuff = try await foo.stuff(context)
        // `LambdaContext` gets this by subtracting now from `.deadline` so
        // we'd need to use delays to test
        if context.getRemainingTime() > 100000 {
            ...
        }
    }
}
```

You can do:

```swift
import LambdaExtrasCore // No AWS dependencies

struct Foo {
    // Doesn't need to know about the context
    var stuff: (String) async throws -> String
}

public struct MyHandler {
    @Dependency(\.foo) var foo
    
    public func run<C>(
        with event: Event,
        context: C
    ) async throws -> String where C: RuntimeContext & EnvironmentValueProvider<Environment> {
        /// Can just access this here and mock
        let bar = try context.value(for: .bar)
        let stuff = try await foo.stuff(bar)
        
        /// Can easily mock this
        if context.getRemainingTime() > 100000 {
            ...
        }
    }
}
```

Reply:
Got it, I'm tracking better now. Thanks!